src/errors.rs

```diff
@@ -0,0 +1,180 @@
+//! Error definitions
+
+pub mod i2c {
+    /// I2C error
+    pub trait Error: core::fmt::Debug {
+        /// Convert error to a generic I2C error kind
+        ///
+        /// By using this method I2C errors freely defined by HAL implementations
+        /// can be converted to a set of generic I2C errors upon which generic
+        /// code can act.
+        fn kind(&self) -> ErrorKind;
+    }
+
+    /// I2C error kind
+    ///
+    /// This represents a common set of I2C operation errors. HAL implementations are
+    /// free to define more specific or additional error types. However, by providing
+    /// a mapping to these common I2C errors, generic code can still react to them.
+    #[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+    #[non_exhaustive]
+    pub enum ErrorKind {
+        /// An unspecific bus error occurred
+        Bus,
+        /// The arbitration was lost, e.g. electrical problems with the clock signal
+        ArbitrationLoss,
+        /// A bus operation was not acknowledged, e.g. due to the addressed device not being available on
+        /// the bus or the device not being ready to process requests at the moment
+        NoAcknowledge,
+        /// The peripheral receive buffer was overrun
+        Overrun,
+        /// The peripheral send buffer ran out of data
+        Underrun,
+        /// A different error occurred. The original error may contain more information.
+        Other,
+    }
+
+    impl Error for ErrorKind {
+        fn kind(&self) -> ErrorKind {
+            *self
+        }
+    }
+
+    impl core::fmt::Display for ErrorKind {
+        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            match self {
+                Self::Bus => write!(f, "An unspecific bus error occurred"),
+                Self::ArbitrationLoss => write!(
+                    f,
+                    "The arbitration was lost, e.g. electrical problems with the clock signal"
+                ),
+                Self::NoAcknowledge => write!(f, "A bus operation was not acknowledged"),
+                Self::Overrun => write!(f, "The peripheral receive buffer was overrun"),
+                Self::Underrun => write!(f, "The peripheral send buffer ran out of data"),
+                Self::Other => write!(
+                    f,
+                    "A different error occurred. The original error may contain more information"
+                ),
```

Comment:
I wonder if those descriptions are a bit _too_ detailed, and should instead go into a `description` method (returning static strings), or something similar.

Reply:
The `description` method was [deprecated](https://doc.rust-lang.org/stable/std/error/trait.Error.html#method.description) in the `std` in favor of the `Display` impl so it would not be good re-adding it here.
An additional drawback would be that with `description` methods, all those strings end up added to the binaries, which causes unnecessary bloat. With the `Display` impl, if you do not use it, it will not be added.
About wording changes, please feel free to suggest any.

Reply:
Ah, thanks, that makes sense. Regarding wording, I'd simply make the descriptions more compact, so instead of "The arbitration was lost, e.g. electrical problems with the clock signal" you could use "Arbitration lost". The more detailed variant could then live in the docs. But maybe other people see value in more detailed error descriptions? I don't know.

Reply:
There is still a difference between `Debug` and `Display`. The `Display` impl is for user-facing output, so I figured a more elaborate formulation would be due, but yeah, "everything is bloat" I guess. I will remove that "e.g." in the arbitration lost message but what do you think about the others?

Reply:
I think the others are fine.

Reply:
I think the others are fine.